### DISCUSSION TITLE
Wrong Power figure when load is high (more than 7000W) on Zmai-90

### DISCUSSION THREAD

**Discussion by tesnaa** [arendst/Tasmota#12114] [Q&A] [answered]
I have zmai-90 successfully flashed with tasmota 9.4.0 (https://templates.blakadder.com/ZMAi-90.html) and been running great so far. I use them to monitor the power usage for the whole house. I just noticed when the load is high, power reported by the tasmota is wrong.

![zmai90 tasmota](https://user-images.githubusercontent.com/80745380/118387144-b83e7900-b646-11eb-9045-fba9275c4535.png)

As you can see the power is only 553.6W when the voltage is 202.4V and Current 35.751A

Checking the debug log also shows similar numbers.

```
06:41:22.349 MQT: tele/tasmota_936F2D/SENSOR = {"Time":"2021-05-16T06:41:22","ENERGY":{"TotalStartTime":"2021-05-14T12:50:32","Total":70.284,"Yesterday":58.880,"Today":11.405,"Period":0.0,"Power":499.7,"ApparentPower":7185.1,"ReactivePower":3238.5,"Factor":0.07,"Voltage":202.5,"Current":35.482}}
```

which shows similar value on Power

The debug also shows outputs the correct power figures on another topic.

```
06:41:11.584 MQT: tele/tasmota_936F2D/RESULT = {"TuyaReceived":{"Data":"55AA030700081302000400011385C3","Cmnd":7,"CmndData":"1302000400011385","DpType2Id19":70533,"19":{"DpId":19,"DpIdType":2,"DpIdData":"00011385"}}}
```

this one has output the correct figure (about 7053.3W) which is the correct value. The zmai also has small LCD display which output this figure.

I have repeat this test consistently, when the load is high somehow the power figures displayed wrong.

As temporary work around, I manually add it to home assistant yaml file to use this mqtt topic instead the default one.

```
platform: mqtt
name: smart_meter_power
state_topic: "tele/tasmota_936F2D/RESULT"
value_template: "{{ (value_json.TuyaReceived.DpType2Id19|float) / 10 }}"
unit_of_measurement: 'W'
```

So any tips how do i fix this issue? what went wrong?

Regards,

Tesna

**Comment by arendst** [comment]
> what went wrong?

Tasmota is tuned to support 240V x 16A max power using uint32_t and floats registers. What happens here is that the values overflow the bounderies of these registers.

There is no easy fix other than totaly rewrite the energy monitoring code to uint64_t and doubles which would greatly increase code size.

So for now it's as designed while someone might manage to solve your tuya specific issue with a niche solution.

**Reply by tesnaa** [reply]
thank you for the reply :) so it's as designed then. Thankfully the correct power measurement data still can be sent to hassio so its not a big deal for me :)

**Reply by barbudor** [reply]
@arendst, may be Tasmota cannot handle more than 3840W in the Energy module.
But still there seems to be a problem in the TuyaMCU module as Power is extracted from the Tuya frame as a `uint16_t`.
I'm expecting the sequence 04 00 01 13 85 to specify a 4-bytes long integer but code reads only lower 2 bytes
```C++
     else if (Tuya.buffer[dpidStart + 1] == 2) {  // Data Type 2
        uint16_t packetValue = Tuya.buffer[dpidStart + 6] << 8 | Tuya.buffer[dpidStart + 7];
...
        } else if (tuya_energy_enabled && fnId == TUYA_MCU_FUNC_POWER) {
          Energy.active_power[0] = (float)packetValue / 10;
          AddLog(LOG_LEVEL_DEBUG, PSTR("TYA: Rx ID=%d Active_Power=%d"), Tuya.buffer[dpidStart], packetValue);
```
Wouldn't it work if `packetValue` was "upgraded" to an `uint32_t` ?

At least re-parsing the value as uint32_t could be done in the if (TUYA_MCU_FUNC_POWER) branch only

**Reply by arendst** [reply]
Give it a try. I haven't looked at the the tuya code, hence my reaction regarding tuya specific ;-)

**Comment by barbudor** [comment, marked as answer]
Can you paste your logs with `WebLog 3` ?
I fear there is a an UINT16 overflow when converting 0x11385 (70533) into Power
With `WebLog 3` You should have logs such as `TYA: RX value %d from dpId %d` and `TYA: Rx ID=%d Active_Power=%d`

Please use the `insert code` formatting for logs readability

**Reply by tesnaa** [reply]
apparently yes its an overflow, as @arendst confirmed. Thankfully the correct power data can stlll be sent to hassio so its not a big deal for me.

**Reply by barbudor** [reply]
@tesnaa Are you compiling your own code ? Or can you do it with a change to test ?

In xdrv_16_tuyamcu.ino around line 858, add the 2nd line as below
```C++
        } else if (tuya_energy_enabled && fnId == TUYA_MCU_FUNC_POWER) {
          uint32_t packetValue = Tuya.buffer[dpidStart + 5] << 16 |Tuya.buffer[dpidStart + 6] << 8 | Tuya.buffer[dpidStart + 7];
          Energy.active_power[0] = (float)packetValue / 10;
          AddLog(LOG_LEVEL_DEBUG, PSTR("TYA: Rx ID=%d Active_Power=%d"), Tuya.buffer[dpidStart], packetValue);
```
That would be extend the extraction of Power by 1 byte without impacting other cases.

Of course, as Theo pointed, some overflow may occur elsewhere.

**Reply by tesnaa** [reply]
@barbudor unfortunately I'm not a coder, so compiling is way out my skills (for now)

 I think it's starting giving out calculation errors in calculation around 65000 mark. Since I also using grafana, the max recorded value is 6.46 kW

![energydemand graph](https://user-images.githubusercontent.com/80745380/118394590-c94fb000-b66f-11eb-9764-4407e0b5f93c.png)


Btw, here is the logs:

```11:42:57.192 MQT: tele/tasmota_936F2D/RESULT = {"TuyaReceived":{"Data":"55AA03070008120200040000808932","Cmnd":7,"CmndData":"1202000400008089","DpType2Id18":32905,"18":{"DpId":18,"DpIdType":2,"DpIdData":"00008089"}}}
11:42:57.202 TYA: fnId=32 is set for dpId=18
11:42:57.205 TYA: RX value 32905 from dpId 18 
11:42:57.207 TYA: Rx ID=18 Current=32905
11:42:57.213 MQT: tele/tasmota_936F2D/RESULT = {"TuyaReceived":{"Data":"55AA0307000813020004000101214D","Cmnd":7,"CmndData":"1302000400010121","DpType2Id19":65825,"19":{"DpId":19,"DpIdType":2,"DpIdData":"00010121"}}}
11:42:57.222 TYA: fnId=31 is set for dpId=19
11:42:57.225 TYA: RX value 289 from dpId 19 
11:42:57.227 TYA: Rx ID=19 Active_Power=289
11:42:57.233 MQT: tele/tasmota_936F2D/RESULT = {"TuyaReceived":{"Data":"55AA0307000814020004000007EE20","Cmnd":7,"CmndData":"14020004000007EE","DpType2Id20":2030,"20":{"DpId":20,"DpIdType":2,"DpIdData":"000007EE"}}}
11:42:57.243 TYA: fnId=33 is set for dpId=20
11:42:57.245 TYA: RX value 2030 from dpId 20 
11:42:57.248 TYA: Rx ID=20 Voltage=2030
11:42:57.911 TYA: Send "55aa00000000ff"
11:42:57.926 MQT: tele/tasmota_936F2D/RESULT = {"TuyaReceived":{"Data":"55AA030000010104","Cmnd":0,"CmndData":"01"}}
11:42:57.935 TYA: Heartbeat
11:43:04.902 MQT: tele/tasmota_936F2D/STATE = {"Time":"2021-05-16T11:43:04","Uptime":"0T23:56:11","UptimeSec":86171,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":52,"POWER":"ON","Wifi":{"AP":1,"SSId":"th0r-IoT","BSSId":"EA:9F:DB:1C:97:3C","Channel":6,"RSSI":78,"Signal":-61,"LinkCount":1,"Downtime":"0T00:00:03"}}
11:43:04.913 MQT: tele/tasmota_936F2D/SENSOR = {"Time":"2021-05-16T11:43:04","ENERGY":{"TotalStartTime":"2021-05-16T09:05:26","Total":2604.811,"Yesterday":58.880,"Today":26.711,"Period":0.1,"Power":28.9,"ApparentPower":6679.7,"ReactivePower":3276.7,"Factor":0.00,"Voltage":203.0,"Current":32.905}}```

**Reply by barbudor** [reply]
Ok, I can build you a `tasmota.bin` with this for testing
Lunch time now, I'll be back later

We can definitively see the problem here:
```
11:42:57.213 MQT: tele/tasmota_936F2D/RESULT = {"TuyaReceived":{"Data":"55AA0307000813020004000101214D","Cmnd":7,"CmndData":"1302000400010121","DpType2Id19":65825,"19":{"DpId":19,"DpIdType":2,"DpIdData":"00010121"}}}
11:42:57.222 TYA: fnId=31 is set for dpId=19
11:42:57.225 TYA: RX value 289 from dpId 19 
11:42:57.227 TYA: Rx ID=19 Active_Power=289
```
Value receive is 65825 (hex : 00 01 01 21) but the value extracted is on only 289 (hex 01 21)

**Reply by barbudor** [reply]
@tesnaa Here attached is a ZIP with both tasmota.bin and tasmota.bin.gz build from yesterday's development version and with the change above.
Please test (keep weblog 3) and report how it works.
[tasmota.zip](https://github.com/arendst/Tasmota/files/6489174/tasmota.zip)

**Reply by tesnaa** [reply]
@barbudor thank you very much for providing me custom build of tasmota. greatly appreciate it. 

It seems your build is working :) now the values goes to more than 65000 :)
![tasmota updated](https://user-images.githubusercontent.com/80745380/118398146-7f6fc580-b681-11eb-9d79-3c0ea96d0916.png)

only now the reactive power shows 0 once the power goes above 6500W. But that's not important for me. Great work! 

logs:
```
13:57:11.725 MQT: tele/tasmota_936F2D/RESULT = {"TuyaReceived":{"Data":"55AA0307000813020004000103E816","Cmnd":7,"CmndData":"13020004000103E8","DpType2Id19":66536,"19":{"DpId":19,"DpIdType":2,"DpIdData":"000103E8"}}}
13:57:11.734 TYA: fnId=31 is set for dpId=19
13:57:11.737 TYA: RX value 1000 from dpId 19 
13:57:11.739 TYA: Rx ID=19 Active_Power=66536
13:57:11.746 MQT: tele/tasmota_936F2D/RESULT = {"TuyaReceived":{"Data":"55AA0307000814020004000007F729","Cmnd":7,"CmndData":"14020004000007F7","DpType2Id20":2039,"20":{"DpId":20,"DpIdType":2,"DpIdData":"000007F7"}}}
13:57:11.756 TYA: fnId=33 is set for dpId=20
13:57:11.758 TYA: RX value 2039 from dpId 20 
13:57:11.761 TYA: Rx ID=20 Voltage=2039
13:57:15.460 MQT: tele/tasmota_936F2D/STATE = {"Time":"2021-05-16T13:57:15","Uptime":"0T00:10:36","UptimeSec":636,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"th0r-IoT","BSSId":"EA:9F:DB:1C:97:3C","Channel":6,"RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:03"}}
13:57:15.473 MQT: tele/tasmota_936F2D/SENSOR = {"Time":"2021-05-16T13:57:15","ENERGY":{"TotalStartTime":"2021-05-16T09:05:26","Total":2610.830,"Yesterday":58.880,"Today":32.730,"Period":20.3,"Power":6653.6,"ApparentPower":6774.6,"ReactivePower":0.0,"Factor":0.98,"Voltage":203.9,"Current":33.225}}
```

**Reply by barbudor** [reply]
@arendst my patch solve the problem at TuyaMCU level

As you expected, some overflow happens in the reactive power calculation which seems to happen here:
https://github.com/arendst/Tasmota/blob/8c53a2d88ba2dc6e087fd0492ca16cb3aa018aef/tasmota/xdrv_03_energy.ino#L981-L990

The overflow occurs when calculating `power * power * 100` > 2^32
I believe the reasons for these uin32/float conversions is to avoid increasing the code size by including the floating point library.

So is my patch in TuyaMCU making sense in a PR considering it's going to fails further in the energy driver ?

**Reply by arendst** [reply]
What happens if you change the uint32_t into uint64_t in line 988 both to result and code increase?

**Reply by barbudor** [reply]
tested using dummy energy driver (after removing the test on values range):

```
16:34:05.934 CMD: backlog voltageset 205.7 ; currentset 32775 ; powerset 6614
16:34:05.947 MQT: stat/nodemcu/VOLTAGESET = {"VoltageSetCal":20570}
16:34:06.197 MQT: stat/nodemcu/CURRENTSET = {"CurrentSetCal":3277500}
16:34:06.446 MQT: stat/nodemcu/POWERSET = {"PowerSetCal":661400}
```

![image](https://user-images.githubusercontent.com/2996042/118401119-b2579080-b664-11eb-8c2f-d90c805bf125.png)

seems ok

before change
RAM:   [======    ]  55.4% (used 45392 bytes from 81920 bytes)
Flash: [======    ]  59.9% (used 613184 bytes from 1023984 bytes)

after change
RAM:   [======    ]  55.4% (used 45392 bytes from 81920 bytes)
Flash: [======    ]  59.9% (used 613296 bytes from 1023984 bytes)

```C++
            reactive_power = (float)(RoundSqrtInt((uint64_t)(apparent_power * apparent_power * 100) - (uint64_t)(Energy.active_power[i] * Energy.active_power[i] * 100))) / 10;
```

**Reply by barbudor** [reply]
Is there many use case where reactive power could be high and overflow ?

Performing the difference in float and converting to uint32 just before the square-root calculation seems to provide the correct values and use less flash than the original code:
```C++
reactive_power = (float)(RoundSqrtInt((uint32_t)((apparent_power * apparent_power - Energy.active_power[i] * Energy.active_power[i]) * 100.0) )) / 10.0;
```
RAM:   [======    ]  55.4% (used 45392 bytes from 81920 bytes)
Flash: [======    ]  59.9% (used 613152 bytes from 1023984 bytes)

**Reply by arendst** [reply]
Well done! The latter is even better. Pls provide PR with both changes.

**Reply by barbudor** [reply]
@tesnaa
When PR https://github.com/arendst/Tasmota/pull/12115 is merged, this will officially be part of the Development branch and available at http://ota.tasmota.com/tasmota

**Reply by tesnaa** [reply]
@barbudor thank you for your help and assistance. greatly appreciate it. Btw,. it seems it has been merged, do I need to wait for couple of hours until the build bot compiles the binary with new patch? or current binaries now has the new patch already?

**Reply by barbudor** [reply]
Sorry for late response
But as a general rule, you can check the time at which the PR is merged and compare to the time of the binaries on ota.tasmota.com. Generally it's build within a few hours.

**Comment by chaveiro** [comment]
This bug is alarming, unexpected power metering above 16A excludes all DIN modules.
Maybe use long for W,  VA, VAr, and KWh with 0.1 precision. (Divided by 10 only before presenting)
Thank you @barbudor for quick bugfixing!
So what are the limits now?
